crates/matrix-sdk/src/client/builder.rs

```diff
@@ -363,71 +395,82 @@ impl ClientBuilder {
 
         let http_client = HttpClient::new(inner_http_client.clone(), self.request_config);
 
-        #[cfg(feature = "experimental-oidc")]
-        let mut authentication_server_info = None;
-        #[cfg(feature = "experimental-oidc")]
-        let allow_insecure_oidc;
+        let (homeserver, well_known) = match homeserver_cfg {
+            HomeserverConfig::Url(url) => (url, None),
 
-        #[cfg(feature = "experimental-sliding-sync")]
-        let mut sliding_sync_proxy: Option<Url> = None;
-
-        let homeserver = match homeserver_cfg {
-            HomeserverConfig::Url(url) => {
-                #[cfg(feature = "experimental-sliding-sync")]
-                {
-                    sliding_sync_proxy =
-                        self.sliding_sync_proxy.as_ref().map(|url| Url::parse(url)).transpose()?;
+            HomeserverConfig::ServerName { server: server_name, protocol } => {
+                let well_known = discover_homeserver(server_name, protocol, &http_client).await?;
+                (well_known.homeserver.base_url.clone(), Some(well_known))
+            }
+
+            HomeserverConfig::ServerNameOrUrl(server_name_or_url) => {
+                // Store the result to return at the end. If this doesn't get modified, then the
+                // supplied name is neither a server name, nor a valid URL.
+                let mut homeserver_details: Option<(
+                    String,
+                    Option<discover_homeserver::Response>,
+                )> = None;
+                let mut discovery_error: Option<ClientBuildError> = None;
+
+                // Attempt discovery as a server name first.
+                let sanitize_result = sanitize_server_name(&server_name_or_url);
+                if let Ok(server_name) = sanitize_result.as_ref() {
+                    let protocol = if server_name_or_url.starts_with("http://") {
+                        UrlScheme::Http
+                    } else {
+                        UrlScheme::Https
+                    };
+
+                    match discover_homeserver(server_name.clone(), protocol, &http_client).await {
+                        Ok(well_known) => {
+                            homeserver_details =
+                                Some((well_known.homeserver.base_url.clone(), Some(well_known)));
+                        }
+                        Err(e) => {
+                            debug!(error = ?e, "Well-known discovery failed.");
```

Comment:
Errors are `Display`able, so you can use:

```suggestion
                            debug!(error = %e, "Well-known discovery failed.");
```